### PR TITLE
Expose prediction functions so we can integrate Workbench with other libraries such as lime.

### PR DESCRIPTION
**Pull request by qimingj**
Updade Scikit-Learn and a few other fixes.

### CODE REVIEW

File: google/datalab/contrib/mlworkbench/_local_predict.py [hunk 1]
```diff
@@ -199,6 +221,51 @@ def get_prediction_results(model_dir_or_id, data, headers, img_cols=None,
...
+  Each instance is ordered by prob. But in some cases probs are needed for fixed
+  order of labels. For example, given labels = ['daisy', 'rose', 'sunflower'], the
+  results of above is expected to be:
+  [
+    [0.8, 0.1, 0.0],
+    [0.01, 0.0, 0.9],
+    ...
+  ]
+  Note that the sum of each instance may not be always 1. If model's top_n is set to
+  none zero, and is less than number of labels, then prediction results may not contain
```
Comment (brandondutra) [inline google/datalab/contrib/mlworkbench/_local_predict.py]: s/'none zero'/'non-zero'

File: google/datalab/ml/_fasets.py [hunk 2]
```diff
@@ -30,6 +31,9 @@ def plot(self, data):
 
     import IPython
 
+    if not isinstance(data, dict) or not all(isinstance(v, pd.DataFrame) for v in data.values()):
+      raise ValueError('Expect a dictionary with values all DataFrames.')
```
Comment (brandondutra) [inline google/datalab/ml/_fasets.py]: 'with values all dataframes' --> where the values are all dataframes

File: google/datalab/contrib/mlworkbench/_local_predict.py [hunk 3]
```diff
@@ -88,9 +94,14 @@ def _download_images(data, img_cols):
   for d in data:
     for img_col in img_cols:
       if d.get(img_col, None):
-        with file_io.FileIO(d[img_col], 'rb') as fi:
-          im = Image.open(fi)
-        images[img_col].append(im)
+        if isinstance(d[img_col], Image.Image):
```
Comment (brandondutra) [inline google/datalab/contrib/mlworkbench/_local_predict.py]: when is this if true, where are images passed to this function?

Comment (qimingj) [inline google/datalab/contrib/mlworkbench/_local_predict.py]: If this if is true, the images are already in "data" in the form of PIL images.

File: google/datalab/contrib/mlworkbench/_local_predict.py [hunk 4]
```diff
@@ -199,6 +221,51 @@ def get_prediction_results(model_dir_or_id, data, headers, img_cols=None,
   return df
 
 
+def get_probs_for_labels(labels, prediction_results):
+  """ Given prediction results, get probs of each label for each instance.
+
+  The prediction results are like:
+  [
+    {'predicted': 'daisy', 'prob': 0.8, 'predicted_2': 'rose', 'prob_2': 0.1},
+    {'predicted': 'sunflower', 'prob': 0.9, 'predicted_2': 'daisy', 'prob_2': 0.01},
```
Comment (brandondutra) [inline google/datalab/contrib/mlworkbench/_local_predict.py]: update 'prob_n' comment to 'probability_'

File: google/datalab/contrib/mlworkbench/_local_predict.py (hunk 3 above)
Comment (brandondutra) [inline google/datalab/contrib/mlworkbench/_local_predict.py]: I know. I think this path will only run if the user does

%%ml predict
prediction_data: $df_with_images

I am worried this is confusing: local prediction data looks different from cloud prediction data.

File: google/datalab/ml/_fasets.py (hunk 2 above)
Comment (qimingj) [inline google/datalab/ml/_fasets.py]: done

File: google/datalab/contrib/mlworkbench/_local_predict.py (hunk 3 above)
Comment (qimingj) [inline google/datalab/contrib/mlworkbench/_local_predict.py]: I added something in get_prediction_results's docstring. Also, I think we will update our notebook sample with images to show how to build a prediction client.

File: google/datalab/contrib/mlworkbench/_local_predict.py (hunk 4 above)
Comment (qimingj) [inline google/datalab/contrib/mlworkbench/_local_predict.py]: done

**Comment by qimingj** [comment]
PTAL

**Pull request merged**